test/page.spec.js

```diff
@@ -1617,4 +1617,183 @@ module.exports.addTests = function({testRunner, expect, puppeteer, DeviceDescrip
+      await page.evaluate(() => {
+        window.__dontFindThis = true;
+        document.location.reload();
+      });
+      expect(await page.evaluate(() => window.__dontFindThis)).not.toBe(true);
+    });
+    it('navigation after a renderer-canceled navigation to a different page', async({page, server}) => {
+      await page.goto(server.EMPTY_PAGE);
+      await page.evaluate(() => {
+        document.location.href = 'grid.html';
+        document.location.href = 'tamperable.html';
+      });
+      expect(await page.evaluate(() => document.location.href)).toBe(server.PREFIX + '/tamperable.html');
+    });
+    it('navigation after a renderer-canceled navigation to the same page', async({page, server}) => {
+      await page.goto(server.EMPTY_PAGE);
+      await page.evaluate(() => {
+        document.location.href = 'grid.html';
+        document.location.href = 'empty.html';
+      });
+      expect(await page.evaluate(() => document.location.href)).toBe(server.EMPTY_PAGE);
+    });
+    it('fail to naviagte', async({page, server}) => {
+      await page.goto(server.EMPTY_PAGE);
+      server.setRoute('/nextpage', async(req, res) => {
+        res.statusCode = 204;
+        res.end();
+      });
+      await page.evaluate(() => {
+        document.location.href = '/nextpage';
+      });
+      expect(await page.evaluate(() => document.location.href)).toBe(server.EMPTY_PAGE);
+    });
+    it('cancel a navigation after it commits and has an execution context', async({page, server}) => {
+      let cleanup;
+      const promise = new Promise(x => cleanup = x);
+      await page.goto(server.EMPTY_PAGE);
+      server.setRoute('/nextpage', async(req, res) => {
+        await evalPromise;
+        res.write('<body><div id="token">token</div>');
+        await page.waitForSelector('#token');
+        await page._client.send('Page.stopLoading');
+        cleanup();
+      });
+      const evalPromise = page.evaluate(() => {
+        document.location.href = '/nextpage';
+      });
+      await evalPromise;
+      expect(await page.evaluate(() => document.location.href)).toBe(server.PREFIX + '/nextpage');
+      await promise;
+      expect(await page.evaluate(() => document.location.href)).toBe(server.PREFIX + '/nextpage');
+    });
+    it('cancel a navigation before it commits', async({page, server}) => {
+      await page.goto(server.EMPTY_PAGE);
+      server.setRoute('/nextpage', async(req, res) => {
+        await page._client.send('Page.stopLoading');
+      });
+      await page.evaluate(() => {
+        document.location.href = '/nextpage';
+      });
+      expect(await page.evaluate(() => document.location.href)).toBe(server.EMPTY_PAGE);
+    });
+    it('cancel a navigation by the renderer after it hits the browser', async({page, server}) => {
+      await page.goto(server.EMPTY_PAGE);
+      server.setRoute('/nextpage', async(req, res) => {
+        await evalPromise;
+        await page.keyboard.down('1');
```

Comment:
can you just evaluate `document.open`?

Reply:
For some reason I can't evaluate in the page while waiting on the initial request, even if I save the possibly stale execution context myself.